### PR TITLE
feat: associations

### PR DESCRIPTION
**Pull request by Hvinder**
### Description
Support associations for all CRMs

Fixes # (issue)

### Type of change
-   [x] New feature (non-breaking change which adds functionality)
-   [x] This change requires a documentation update

### How Has This Been Tested?
-   [x] Tested with postman, and added relevant examples

### Checklist:

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my code
-   [ ] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have made corresponding changes to the documentation
-   [x] My changes generate no new warnings
-   [ ] I have added tests that prove my fix is effective or that my feature works
-   [ ] New and existing unit tests pass locally with my changes
-   [x] Any dependent changes have been merged and published in downstream modules

### Associations checklist
(❌ -> not supported)
Hubspot
<table>
    <tr>
        <th></th>
        <th>leadId</th>
        <th>dealId</th>
        <th>contactId</th>
        <th>companyId</th>
        <th>noteId</th>
        <th>eventId</th>
        <th>taskId</th>
    </tr>
    <tr>
        <td>lead</td>
        <td>NA</td>
        <td>✅</td>
        <td>❌</td>
        <td>✅</td>
        <td></td>
        <td></td>
        <td></td>
    </tr>
    <tr>
        <td>deal</td>
        <td></td>
        <td>NA</td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
    </tr>    
    <tr>
        <td>contact</td>
        <td>❌</td>
        <td>✅</td>
        <td>NA</td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
    </tr>    
    <tr>
        <td>company</td>
        <td></td>
        <td>✅</td>
        <td></td>
        <td>NA</td>
        <td></td>
        <td></td>
        <td></td>
    </tr>
    <tr>
        <td>note</td>
        <td>✅</td>
        <td>✅</td>
        <td>✅</td>
        <td>✅</td>
        <td>NA</td>
        <td></td>
        <td></td>
    </tr>
    <tr>
        <td>event</td>
        <td></td>
        <td>✅</td>
        <td>✅</td>
        <td></td>
        <td>❌</td>
        <td>NA</td>
        <td></td>
    </tr>
    <tr>
        <td>task</td>
        <td>❌</td>
        <td>✅</td>
        <td></td>
        <td></td>
        <td>❌</td>
        <td></td>
        <td>NA</td>
    </tr>
</table>

Zoho
<table>
    <tr>
        <th></th>
        <th>leadId</th>
        <th>dealId</th>
        <th>contactId</th>
        <th>companyId</th>
        <th>noteId</th>
        <th>eventId</th>
        <th>taskId</th>
    </tr>
    <tr>
        <td>lead</td>
        <td>NA</td>
        <td>❌</td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
    </tr>
    <tr>
        <td>deal</td>
        <td></td>
        <td>NA</td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
    </tr>    
    <tr>
        <td>contact</td>
        <td></td>
        <td>✅</td>
        <td>NA</td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
    </tr>    
    <tr>
        <td>company</td>
        <td></td>
        <td>✅</td>
        <td></td>
        <td>NA</td>
        <td></td>
        <td></td>
        <td></td>
    </tr>
    <tr>
        <td>note</td>
        <td></td>
        <td>✅</td>
        <td></td>
        <td></td>
        <td>NA</td>
        <td></td>
        <td></td>
    </tr>
    <tr>
        <td>event</td>
        <td></td>
        <td>✅</td>
        <td></td>
        <td></td>
        <td></td>
        <td>NA</td>
        <td></td>
    </tr>
    <tr>
        <td>task</td>
        <td></td>
        <td>✅</td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td>NA</td>
    </tr>
</table>

Salesforce
<table>
    <tr>
        <th></th>
        <th>leadId</th>
        <th>dealId</th>
        <th>contactId</th>
        <th>companyId</th>
        <th>noteId</th>
        <th>eventId</th>
        <th>taskId</th>
    </tr>
    <tr>
        <td>lead</td>
        <td>NA</td>
        <td>❌</td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
    </tr>
    <tr>
        <td>deal</td>
        <td></td>
        <td>NA</td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
    </tr>    
    <tr>
        <td>contact</td>
        <td></td>
        <td>✅</td>
        <td>NA</td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
    </tr>    
    <tr>
        <td>company</td>
        <td></td>
        <td>✅</td>
        <td></td>
        <td>NA</td>
        <td></td>
        <td></td>
        <td></td>
    </tr>
    <tr>
        <td>note</td>
        <td></td>
        <td>✅</td>
        <td></td>
        <td></td>
        <td>NA</td>
        <td></td>
        <td></td>
    </tr>
    <tr>
        <td>event</td>
        <td></td>
        <td>✅</td>
        <td>✅</td>
        <td></td>
        <td></td>
        <td>NA</td>
        <td></td>
    </tr>
    <tr>
        <td>task</td>
        <td></td>
        <td>✅</td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td>NA</td>
    </tr>
</table>

Pipedrive
<table>
    <tr>
        <th></th>
        <th>leadId</th>
        <th>dealId</th>
        <th>contactId</th>
        <th>companyId</th>
        <th>noteId</th>
        <th>eventId</th>
        <th>taskId</th>
    </tr>
    <tr>
        <td>lead</td>
        <td>NA</td>
        <td></td>
        <td>✅</td>
        <td>✅</td>
        <td></td>
        <td></td>
        <td></td>
    </tr>
    <tr>
        <td>deal</td>
        <td></td>
        <td>NA</td>
        <td>✅</td>
        <td>✅</td>
        <td></td>
        <td></td>
        <td></td>
    </tr>    
    <tr>
        <td>contact</td>
        <td></td>
        <td></td>
        <td>NA</td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
    </tr>    
    <tr>
        <td>company</td>
        <td></td>
        <td></td>
        <td></td>
        <td>NA</td>
        <td></td>
        <td></td>
        <td></td>
    </tr>
    <tr>
        <td>note</td>
        <td>✅</td>
        <td>✅</td>
        <td>✅</td>
        <td>✅</td>
        <td>NA</td>
        <td></td>
        <td></td>
    </tr>
    <tr>
        <td>event</td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td>NA</td>
        <td></td>
    </tr>
    <tr>
        <td>task</td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td>NA</td>
    </tr>
</table>

### CODE REVIEW

File: packages/backend/services/crm/contact.ts [hunk 1]
```diff
@@ -225,6 +227,16 @@ const contactService = new ContactService(
                 const contact = disunifyContact(contactData, thirdPartyId);
                 console.log('Revert::CREATE CONTACT', tenantId, contact, thirdPartyId);
 
+                if (
```
Comment (jatinsandilya) [inline packages/backend/services/crm/contact.ts:230]: @Hvinder Lets have this logic in the zoho block so things are scoped properly?

File: packages/backend/models/unified/note.ts [hunk 2]
```diff
@@ -36,14 +38,14 @@ export function unifyNote(note: any, tpId: TP_ID): UnifiedNote {
...
-                organizationId: note.org_id,
+                contactId: note.person_id,
+                companyId: note.org_id,
                 leadId: note.lead_id,
                 dealId: note.deal_id,
             }),
-            ...(tpId === TP_ID.hubspot && {
-                deal_id: '',
-            }),
+            // ...(tpId === TP_ID.hubspot && {
```
Comment (jatinsandilya) [inline packages/backend/models/unified/note.ts:46]: @Hvinder Is this commented out for a reason?

File: packages/backend/helpers/hubspot.ts [hunk 3]
```diff
@@ -1,16 +1,68 @@
...
-        }
+export const getHubspotAssociationObj = (
+    key: AllAssociation,
+    associateObj: 'note' | 'deal' | 'contact' | 'lead' | 'company' | 'event' | 'task'
+) => {
+    const associationTypeMapping: {
+        [x in typeof associateObj]: { [y in AllAssociation]: number | undefined };
+    } = {
+        note: {
+            dealId: 214,
```
Comment (jatinsandilya) [inline packages/backend/helpers/hubspot.ts:11]: @Hvinder Is this ok to be hardcoded?

File: packages/backend/models/unified/note.ts (hunk 2 above)
Comment (Hvinder) [reply packages/backend/models/unified/note.ts:46]: yeah, we don't need that as we don't get the associated `deal_id` in the response

**Review by jatinsandilya — APPROVED** [review]
LGTM 🏃🏼

**Pull request merged**